plugins/transformations/seq2array/srcCxx/Seq2ArrayTransformation.cxx

```diff
@@ -0,0 +1,481 @@
+ * works of the Software.  Licensee has the right to distribute object form
+ * only for use with RTI products.  The Software is provided "as is", with no
+ * warranty of any type, including any warranty for fitness for any purpose.
+ * RTI is under no obligation to maintain or support the Software.  RTI shall
+ * not be liable for any incidental or consequential damages arising out of the
+ * use or inability to use the software.
+ */
+#include <dds/dds.hpp>
+#include <rti/routing/RoutingService.hpp>
+#include <rti/routing/transf/TransformationPlugin.hpp>
+#include <rti/core/xtypes/DynamicDataImpl.hpp>
+
+#include "Seq2ArrayTransformation.hpp"
+#include "DynamicDataHelpers.hpp"
+
+using namespace dds::core::xtypes;
+using dds::sub::SampleInfo;
+
+/*
+ * @brief Check that both types are compatible, a compatible type for this
+ * transformation will mean that all the elements are the same but the sequences
+ * in the input are replaced by arrays in the output.
+ *
+ */
+bool Seq2ArrayTransformation::are_type_compatible(
+        const dds::core::xtypes::DynamicType & input_type,
+        const dds::core::xtypes::DynamicType & output_type)
+{
+    bool is_compatible = true;
+
+    switch (input_type.kind().underlying()) {
+    case TypeKind::STRUCTURE_TYPE: {
+        if (output_type.kind() != TypeKind::STRUCTURE_TYPE) {
+            std::string error("Error: the ouput field <" + output_type.name()
+                    + "> is different to the input type <" + input_type.name());
+            throw std::runtime_error(error);
+        }
+
+        auto input_struct_type = static_cast<const StructType &>(input_type);
+        auto output_struct_type = static_cast<const StructType &>(output_type);
+
+        if (input_struct_type.member_count() != output_struct_type.member_count()) {
+            std::string error("Error: different member count of input and output "
+                    "struct, input name: <" + input_struct_type.name()
+                    + ">, output name <" + output_struct_type.name() + ">: "
+                    + std::to_string(input_struct_type.member_count())
+                    + " != " + std::to_string(output_struct_type.member_count()));
+            throw std::runtime_error(error);
+        }
+
+        if (input_struct_type == output_struct_type) {
+            return true;
+        } else {
+            // inspect element by element since it may be compatible
+            // (output with arrays and input with sequences)
+            for (auto member : input_struct_type.members()) {
+                switch (member.type().kind().underlying()) {
+                case TypeKind::STRUCTURE_TYPE:
+                case TypeKind::SEQUENCE_TYPE:
+                case TypeKind::ARRAY_TYPE:
+                case TypeKind::UNION_TYPE:
+                    is_compatible = is_compatible && are_type_compatible(
+                            member.type(),
+                            output_struct_type.member(member.name()).type());
```

Comment:
the main difference is that the member ID (accessible from the DynamicType) and the index (accessible from the DynamicData) is that the member ID identifies the member and may not be consecutive, and the index is the order in the DynamicData (consecutive).
Then, the type compatibility check is done with the member name (the other option would be the member ID), and when accessing the DynamicData, the transformation does it element by element (using the index). Therefore, both DynamicData must contain the same elements in the same position (just changing the sequences to arrays). 

I'll add this information to the doc.

About the error, if the output type doesn't have that member, it will throw an exception, which makes sense.

And about the `output_struct_type.member(member.name())`, actually, in the "success" codepath, it is just called once, in the error codepath it is called twice. I've moved it outside the `switch` statement.

Reply:
BTW, before moving the checking of the type to a separate function, it was only relying on the `index`, then when I created this function (are_types_compatible) I added the checking with the names. So you were not missing anything, I changed the behavior but not the doc.